src/v/cluster/rm_stm.cc

```diff
@@ -1009,47 +1012,184 @@ ss::future<result<raft::replicate_result>> rm_stm::replicate_seq(
         co_return errc::not_leader;
     }
     auto synced_term = _insync_term;
-
+    
     if (bid.first_seq > bid.last_seq) {
-        vlog(
-          clusterlog.warn,
-          "first_seq={} of the batch should be less or equal to last_seq={}",
-          bid.first_seq,
-          bid.last_seq);
-        outcome->set_value();
+        vlog(clusterlog.warn, "first_seq={} of the batch should be less or equal to last_seq={}", bid.first_seq, bid.last_seq);
         co_return errc::generic_tx_error;
     }
 
+    // getting session by client's pid
+    ss::lw_shared_ptr<inflight_requests> session;
+    auto session_it = _inflight_requests.find(bid.pid);
+    if (session_it == _inflight_requests.end()) {
+        session = ss::make_lw_shared<inflight_requests>();
+        _inflight_requests.try_emplace(bid.pid, session);
+    } else {
+        session = session_it->second;
+    }
+
+    // critical section, rm_stm evaluates the request its order and
+    // whether or not it's duplicated
+    // ASSUMING THAT ALL INFLIGHT REPLICATION REQUESTS WILL SUCCEED
+    // if any of the request fail we expect that the leader step down
+    auto u = co_await session->lock.get_units();
+
+    if (!_c->is_leader() || _c->term() != synced_term) {
+        u.return_all();
+        co_return errc::not_leader;
+    }
+
+    if (session->term < synced_term) {
+        // we don't care about old inflight requests because the sync
+        // guarantee (all replicated records of the last term are
+        // applied) makes sure that the passed inflight records got
+        // into _log_state->seq_table
+        session->forget();
+        session->term = synced_term;
+    }
+    
+    // checking if the request (identified by seq) is already resolved
+    // checking among the pending requests
+    auto cached_r = session->known_seq(bid.last_seq);
+    if (cached_r) {
+        u.return_all();
+        co_return cached_r.value();
+    }
+    // checking among the responded requests
     auto cached_offset = known_seq(bid);
     if (cached_offset) {
-        if (cached_offset.value() < model::offset{0}) {
-            vlog(
-              clusterlog.warn,
-              "Status of the original attempt is unknown (still is in-flight "
-              "or failed). Failing a retried batch with the same pid:{} & "
-              "seq:{} to avoid duplicates",
-              bid.pid,
-              bid.last_seq);
-            // kafka clients don't automaticly handle fatal errors
-            // so when they encounter the error they will be forced
-            // to propagate it to the app layer
-            co_return errc::generic_tx_error;
-        }
+        u.return_all();
         co_return raft::replicate_result{.last_offset = cached_offset.value()};
     }
 
-    if (!check_seq(bid)) {
-        co_return errc::sequence_out_of_order;
+    // checking if the request is already being processed
+    for (auto inflight : session->cache) {
```

Comment:
this can be `std::find_if`

Reply:
I've tried it but it too noisy (visual noise, more code and an extra variable in the scope):

```
auto inflight_it = std::find_if(
    session->cache.cbegin(),
    session->cache.cend(),
    [](ss::lw_shared_ptr<inflight_request> inflight) {
        return inflight->last_seq == bid.last_seq && inflight->is_processing;
    });

if (inflight_it != session->cache.cend()) {
    // found an inflight request, parking the current request
    // until the former is resolved
    auto promise = ss::make_lw_shared<available_promise<result<raft::replicate_result>>>();
    inflight->parked.push_back(promise);
    u.return_all();
    co_return co_await promise->get_future();
}
```

The original approach looks simpler:

```
for (auto inflight : session->cache) {
    if (inflight->last_seq == bid.last_seq && inflight->is_processing) {
        // found an inflight request, parking the current request
        // until the former is resolved
        auto promise = ss::make_lw_shared<available_promise<result<raft::replicate_result>>>();
        inflight->parked.push_back(promise);
        u.return_all();
        co_return co_await promise->get_future();
    }
}
```

Sticking with it